src/odemis/acq/align/test/fastem_test.py

```diff
@@ -0,0 +1,180 @@
+Odemis. If not, see http://www.gnu.org/licenses/.
+
+"""
+from __future__ import division
+
+import logging
+import os
+import time
+import unittest
+from concurrent.futures._base import CancelledError
+
+import numpy
+
+import odemis
+from odemis import model
+from odemis.acq.align import fastem
+from odemis.acq.align.fastem import OPTICAL_AUTOFOCUS, IMAGE_TRANSLATION_PREALIGN
+from odemis.util import test
+
+# * TEST_NOHW = 1: use simulator (asm/sam and xt adapter simulators need to be running)
+# * TEST_NOHW = 0: connected to the real hardware (backend needs to be running)
+# technolution_asm_simulator/simulator2/run_the_simulator.py
+TEST_NOHW = (os.environ.get("TEST_NOHW", "0") != "0")  # Default is HW testing
+
+logging.getLogger().setLevel(logging.DEBUG)
+logging.basicConfig(format="%(asctime)s  %(levelname)-7s %(module)s:%(lineno)d %(message)s")
+
+CONFIG_PATH = os.path.dirname(odemis.__file__) + "/../../install/linux/usr/share/odemis/"
+FASTEM_CONFIG = CONFIG_PATH + "sim/fastem-sim-asm.odm.yaml"
+
+
+class TestFastEMCalibration(unittest.TestCase):
+    """Test the calibration manager."""
+
+    @classmethod
+    def setUpClass(cls):
+        if TEST_NOHW:
+            test.start_backend(FASTEM_CONFIG)
+
+        # get the hardware components
+        cls.scanner = model.getComponent(role='e-beam')
+        cls.asm = model.getComponent(role="asm")
+        cls.mppc = model.getComponent(role="mppc")
+        cls.multibeam = model.getComponent(role="multibeam")
+        cls.descanner = model.getComponent(role="descanner")
+        cls.stage = model.getComponent(
+            role="stage")  # TODO replace with stage-scan when ROA conversion method available
+        cls.ccd = model.getComponent(role="diagnostic-ccd")
+        cls.focuser = model.getComponent(role="diagnostic-cam-focus")
+        cls.beamshift = model.getComponent(role="ebeam-shift")
+        cls.det_rotator = model.getComponent(role="det-rotator")
+
+    def setUp(self):
+
+        self.good_focus = -70e-6  # position where the image of the multiprobe is displayed in focus [m]
+        self.ccd.updateMetadata(model.MD_FAV_POS_ACTIVE.update({"z": self.good_focus}))
+
+        # move the stage so that the image is in focus
+        self.focuser.moveAbs({"z": self.good_focus}).result()
+
+    def test_optical_autofocus(self):
+        """Run the optical autofocus calibration. Can also be tested with simulator."""
+
+        calibrations = [OPTICAL_AUTOFOCUS]
+
+        # move the stage so that the image is out of focus
+        self.center_position = -30e-6
+        self.focuser.moveAbs({"z": self.center_position}).result()
+
+        # Run auto focus
+        f = fastem.align(self.scanner, self.multibeam, self.descanner, self.mppc, self.stage, self.ccd,
+                         self.beamshift, self.det_rotator, calibrations)
+
+        # foc_pos, foc_lev = f.result(timeout=900)  # TODO make use of the focus level?
```

Comment:
you could add an assert that the focus level is larger than 1e10, at least for the hardware it should be something like that, I don't know what to expect for the simulator.

Reply:
I just noticed, that we do not return the focus level and focus position yet. The future only returns exceptions in the calibration manager and nothing else. I don't know if we can extend the calibration manager to return such values (and only for the autofocus then). We do return DataArrays for the acquisition. But those are captured via a threading event I think.

Reply:
I think for now let's not do this?